assets/sip-big_vector/big_vector.move

```diff
@@ -0,0 +1,339 @@
+    public fun is_empty<Element: store>(bv: &BigVector<Element>): bool {
+        bv.length == 0
+    }
+
+    /// push a new element at the end of the BigVector
+    public fun push_back<Element: store>(bv: &mut BigVector<Element>, element: Element) {
+        if (length(bv) / bv.slice_size == bv.slice_count) {
+            bv.slice_count = bv.slice_count + 1;
+            let new_slice = vector::singleton(element);
+            dynamic_field::add(&mut bv.id, bv.slice_count, new_slice);
+        }
+        else {
+            let slice = dynamic_field::borrow_mut(&mut bv.id, bv.slice_count);
+            vector::push_back(slice, element);
+        };
+        bv.length = bv.length + 1;
+    }
+
+    /// pop an element from the end of the BigVector
+    public fun pop_back<Element: store>(bv: &mut BigVector<Element>): Element {
+        let slice = dynamic_field::borrow_mut(&mut bv.id, bv.slice_count);
+        let element = vector::pop_back(slice);
+        trim_slice(bv);
+        bv.length = bv.length - 1;
+
+        element
+    }
+
+    /// borrow an element at index i from the BigVector
+    public fun borrow<Element: store>(bv: &BigVector<Element>, i: u64): &Element {
+        let slice_count = (i / bv.slice_size) + 1;
+        let slice = dynamic_field::borrow(&bv.id, slice_count);
+        vector::borrow(slice, i % bv.slice_size)
+    }
+
+    /// borrow a mutable element at index i from the BigVector
+    public fun borrow_mut<Element: store>(bv: &mut BigVector<Element>, i: u64): &mut Element {
+        let slice_count = (i / bv.slice_size) + 1;
+        let slice = dynamic_field::borrow_mut(&mut bv.id, slice_count);
+        vector::borrow_mut(slice, i % bv.slice_size)
+    }
+
+    /// borrow a slice from the BigVector
+    public fun borrow_slice<Element: store>(bv: &BigVector<Element>, slice_count: u64): &vector<Element> {
+        dynamic_field::borrow(&bv.id, slice_count)
+    }
+
+    /// borrow a mutable slice from the BigVector
+    public fun borrow_slice_mut<Element: store>(bv: &mut BigVector<Element>, slice_count: u64): &mut vector<Element> {
+        dynamic_field::borrow_mut(&mut bv.id, slice_count)
+    }
+
+    /// swap and pop the element at index i with the last element
+    public fun swap_remove<Element: store>(bv: &mut BigVector<Element>, i: u64): Element {
+        let result = pop_back(bv);
+        if (i == length(bv)) {
+            result
+        } else {
+            let slice_count = (i / bv.slice_size) + 1;
+            let slice = dynamic_field::borrow_mut<u64, vector<Element>>(&mut bv.id, slice_count);
+            vector::push_back(slice, result);
+            vector::swap_remove(slice, i % bv.slice_size)
+        }
+    }
+
+    /// remove the element at index i and shift the rest elements
+    /// abort when reference more thant 1000 slices
+    /// costly function, use wisely
+    public fun remove<Element: store>(bv: &mut BigVector<Element>, i: u64): Element {
```

Comment:
I am not sure about this API and I keep flipping between feeling ok with it and not.
I find it difficult to relate to a framework API that has a chance of failing. 
I understand that is "life" but what looks an innocuous call that can be heavily tested and should always succeed but then fail may be too much.
In other words if I write a contract that uses big vector and rely on this API for some critical functionality I now have the possibility that the contract is made unusable by a failure in this call.
For instance, say a contract has an API that gives my asset back and I was the first asset in the vector that got over 1K slices. Now I have no way to get my asset back.
If internal logic relies on this it may be possible the entire contract gets broken.
I am inclined to remove this though I understand it could be useful in many scenarios.

My current inclination is to remove this API and make clear that all removal in big vector are not stable.

The alternative would be to have slice size variable. That obviously adds complexity and it may be out of scope for this SIP, but it may also be the only good way to have a `remove` API that is sound.

Curious to hear other people thoughts

Reply:
I need to do some more thinking about this, but the path this leads me down (if we definitely want to support `remove`) is to opt for a structure that's somewhere between a B-Tree and a Rope.

Reply:
I think this function should rename to `remove_unsafe`, and maybe provide other solution to achieve it. In Typus v1, we split a transaction which accesses more than 1000 dynamic_fields into several transactions in order to bypass the limitation. So, there may be a function named `remove_element` which only removes an element from BigVector and lock the BigVector, and then a function such as `public fun comb<E: store>(self: &mut BigVector<E>, counter: u16)` that shifting all subsequent elements. The counter should less than 1000, and unlock the BigVector after several transactions until last element is processed.